src/engine/ExecuteUpdate.cpp

```diff
@@ -46,61 +46,101 @@ std::pair<std::vector<ExecuteUpdate::TransformedTriple>, LocalVocab>
 ExecuteUpdate::transformTriplesTemplate(
     const EncodedIriManager& encodedIriManager, const Index::Vocab& vocab,
     const VariableToColumnMap& variableColumns,
-    std::vector<SparqlTripleSimpleWithGraph>&& triples) {
+    const std::vector<SparqlTripleSimpleWithGraph>& triples) {
+  // - collect all TripleComponents that live in the Vocabulary (literals and
+  // IRIs)
+  // - deduplicate and sort them according to the order in the Vocabulary
+  // - create a map with the TC and their respective Id
+  // - transform the triples using the map and the variableColumns to
+  // `IdOrVariableIndex`
+  // We do this to access the disk in an optimal pattern. The vocabulary does a
+  // scan with jumps over the disk for lookups instead of random accesses.
+  ad_utility::HashSet<TripleComponent> lookupItems;
+  // Collect the Literals and Iri for the Vocab lookup.
+  auto addTCToLookup = [&lookupItems](const TripleComponent& tc) {
+    if (tc.isLiteral() || tc.isIri()) {
+      lookupItems.insert(tc);
+    }
+  };
+  auto addGraphToLookup =
+      [&lookupItems](const SparqlTripleSimpleWithGraph::Graph& g) {
+        // Ignore the default graph (std::monostate) here, because it is always
+        // added.
+        if (const auto* iri = std::get_if<TripleComponent::Iri>(&g)) {
+          lookupItems.insert(TripleComponent(*iri));
+        }
+      };
+  for (const auto& triple : triples) {
+    addTCToLookup(triple.s_);
+    addTCToLookup(triple.p_);
+    addTCToLookup(triple.o_);
+    addGraphToLookup(triple.g_);
+  }
+  // Always adding the default graph is no overhead and we need it for
+  // `defaultGraphIri`.
+  lookupItems.insert(TripleComponent(
+      ad_utility::triple_component::Iri::fromIriref(DEFAULT_GRAPH_IRI)));
+  std::vector lookupVec(lookupItems.begin(), lookupItems.end());
```

Comment:
```suggestion
  std::vector lookupVec(std::move_iterator(lookupItems.begin()), std::move_iterator(lookupItems.end()));
```

This should hopefully avoid expensive copies.